### PR TITLE
Add a SetParameter action that sets a parameter to all nodes in the same scope

### PR DESCRIPTION
**Pull request by ivanpauno**
Similarly to how `PushRosNamespace` action works, this allows a way to set a parameter affecting all `Node` actions and `ComposableNode` descriptions in the same scope.

This partially fixes https://github.com/ros2/launch_ros/issues/45, the other half is doing the same for remapping rules.

This introduces some conflicts with https://github.com/ros2/launch_ros/pull/137, but either way I already have to rebase that PR.

### CODE REVIEW

File: launch_ros/launch_ros/actions/push_ros_namespace.py [hunk 1]
```diff
@@ -29,7 +29,8 @@
 
 from rclpy.validate_namespace import validate_namespace
 
-
+# TODO(ivanpauno): Generalize launch.frontend so both hyphens and underscores are equivalent.
+@expose_action('push_ros_namespace')
```
Comment (hidmic) [inline launch_ros/launch_ros/actions/push_ros_namespace.py:33]: @ivanpauno should this be in its own separate PR? Same for all the other changes in this file.

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+)
+def test_set_param(params_to_set, expected_result):
+    set_parameter_actions = []
+    for dicts in params_to_set:
+        for name, value in dicts.items():
+            set_parameter_actions.append(SetParameter(name=name, value=value))
+    lc = MockContext()
+    for set_param in set_parameter_actions:
+        set_param.execute(lc)
+    lc.launch_configurations == {'ros_params': expected_result}
```
Comment (hidmic) [inline test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py:62]: @ivanpauno
```suggestion
    assert lc.launch_configurations == {'ros_params': expected_result}
```
?

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        namespace='my_ns',
+        parameters=[{'asd': 'bsd'}]
+    )
+    set_param_1 = SetParameter(name='my_param', value='my_value')
+    set_param_2 = SetParameter(name='asd', value='csd')
+    set_param_1.execute(lc)
+    set_param_2.execute(lc)
+    request = get_composable_node_load_request(node_description, lc)
+    parameters = request.parameters
+    print(parameters)
```
Comment (hidmic) [inline test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py:110]: @ivanpauno a leftover?

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            id='Two params, overriding one'
+        ),
+    ]
+
+
+@pytest.mark.parametrize(
+    'params_to_set, expected_result',
+    get_set_parameter_test_parameters()
+)
+def test_set_param(params_to_set, expected_result):
```
Comment (hidmic) [inline test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py:54]: @ivanpauno nit: it'd be nice to include a test to verify scoping works e.g. "global" parameters within a group are not visible outside of it.

File: launch_ros/launch_ros/actions/node.py [hunk 5]
```diff
@@ -385,11 +379,21 @@ def _perform_substitutions(self, context: LaunchContext) -> None:
...
+        # so they can be overriden with specific parameters of this Node
+        global_params = context.launch_configurations.get('ros_params', None)
+        if global_params is not None or self.__parameters is not None:
+            self.__expanded_parameter_files = []
+        if global_params is not None:
+            param_file_path = self._create_params_file_from_dict(global_params)
+            self.__expanded_parameter_files.append(param_file_path)
+            cmd_extension = ['--params-file', f'{param_file_path}']
+            self.cmd.extend([normalize_to_list_of_substitutions(x) for x in cmd_extension])
+            assert os.path.isfile(param_file_path)
```
Comment (hidmic) [inline launch_ros/launch_ros/actions/node.py:392]: @ivanpauno here we `assert`, but below we may ignore a missing file under the same circumstances. I think we should be consistent.

File: launch_ros/launch_ros/actions/load_composable_nodes.py [hunk 6]
```diff
@@ -208,3 +172,53 @@ def execute(
...
+    if composable_node_description.remappings is not None:
+        for from_, to in composable_node_description.remappings:
+            request.remap_rules.append('{}:={}'.format(
+                perform_substitutions(context, list(from_)),
+                perform_substitutions(context, list(to)),
+            ))
+    global_params = context.launch_configurations.get('ros_params', None)
+    parameters = composable_node_description.parameters
+    if global_params is not None:
+        parameters = (normalize_parameter_dict(global_params), *parameters)
```
Comment (hidmic) [inline launch_ros/launch_ros/actions/load_composable_nodes.py:207]: @ivanpauno if `parameters` may be `None`, iterable expansion may fail.

File: launch_ros/launch_ros/actions/load_composable_nodes.py [hunk 7]
```diff
@@ -208,3 +172,53 @@ def execute(
...
+    if composable_node_description.parameters is not None:
```
Comment (hidmic) [inline launch_ros/launch_ros/actions/load_composable_nodes.py:208]: @ivanpauno this should occur if only global parameters are available too.

File: launch_ros/launch_ros/actions/node.py (hunk 5 above)
Comment (ivanpauno) [reply launch_ros/launch_ros/actions/node.py:392]: it's not exactly the same, this is a temporal file created a few lines before.

The other might be a temporal file or a file passed by the user.
I can add an assert for the temp file, but it's not the goal of this PR to change what happens when the user passes a wrong path.

File: launch_ros/launch_ros/actions/push_ros_namespace.py (hunk 1 above)
Comment (ivanpauno) [reply launch_ros/launch_ros/actions/push_ros_namespace.py]: Reverted in https://github.com/ros2/launch_ros/pull/158/commits/f66375545bf968f686c60e94fbc961f853b5109c. I will do this in a follow-up PR.

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py (hunk 2 above)
Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py]: :joy:
fixed in https://github.com/ros2/launch_ros/pull/158/commits/f66375545bf968f686c60e94fbc961f853b5109c

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py (hunk 3 above)
Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py]: yeap, deleted in https://github.com/ros2/launch_ros/pull/158/commits/f66375545bf968f686c60e94fbc961f853b5109c

File: launch_ros/launch_ros/actions/node.py (hunk 5 above)
Comment (ivanpauno) [reply launch_ros/launch_ros/actions/node.py:392]: done https://github.com/ros2/launch_ros/pull/158/commits/f66375545bf968f686c60e94fbc961f853b5109c

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 6 above)
Comment (ivanpauno) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: fixed https://github.com/ros2/launch_ros/pull/158/commits/f66375545bf968f686c60e94fbc961f853b5109c

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 7 above)
Comment (ivanpauno) [reply launch_ros/launch_ros/actions/load_composable_nodes.py:208]: fixed https://github.com/ros2/launch_ros/pull/158/commits/f66375545bf968f686c60e94fbc961f853b5109c

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py (hunk 4 above)
Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py:58]: Added in https://github.com/ros2/launch_ros/pull/158/commits/e7c780c41d9b7ce19bdc8246da882e591b682db0

**Comment by ivanpauno** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11328/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6566/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9292/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11224/)

**Comment by wjwwood** [comment]
(I had already started the review, so I finished it)
